ale/teacher/exploitation/fluctuation_historical_sequence.py

```diff
@@ -110,68 +103,76 @@ def __init__(
             aggregation_method=aggregation_method
         )
         random.seed(self.seed)
-        self.historical_sequence = None
-        self.historical_window_size = historical_window_size
-        self.weight = weight
+        self.historical_sequence: Optional[HistoricalSequence] = None
+        self.historical_window_size: int = historical_window_size
+        self.weight: float = weight
 
     def propose(self, potential_ids: List[int], step_size: int, budget: int) -> List[int]:
         search_for_least_confidence = random.sample(potential_ids, budget)
         idx2text = self.corpus.get_text_by_ids(search_for_least_confidence)
-        prediction_results: Dict[int, PredictionResult] = self.predictor.predict(
-            idx2text)
-        out_ids: List[int] = self.compute_function(
-            prediction_results, step_size)
+        prediction_results: Dict[int, PredictionResult] = self.predictor.predict(idx2text)
+        out_ids: List[int] = self.compute_function(prediction_results, step_size)
+
+        return out_ids
+
+    def compute_ner(self, predictions: Dict[int, PredictionResult], step_size: int) -> List[int]:
+        """
+        Fluctuation of Historical Sequence is calculated on instance-level,
+        LC is used with token-score, aggregated as configured.
+        """
+        if self.is_initial_train():  # TODO Question: Do we have a model at this point?
+            next_batch = random.sample(list(predictions.keys()), step_size)
+            instance_scores: Dict[int, float] = self.compute_least_confidence_instance_scores(predictions)
+            self.historical_sequence: HistoricalSequence = HistoricalSequence(instance_scores,
+                                                                              self.historical_window_size)
+            return next_batch  # TODO Question: returning the random batch for comparison reason?
+
+        instance_scores: Dict[int, float] = self.compute_least_confidence_instance_scores(predictions)
+        historical_sequences: Dict[int, List[float]] = self.historical_sequence.update_and_get_historical_sequences(
+            instance_scores)
+
+        fluctuation_scores: Dict[int, float] = self.compute_fluctuation_scores_per_doc(historical_sequences)
+
+        # LC calculated with 1-highest score, therefore take max
+        sorted_dict_by_score = sorted(fluctuation_scores.items(), key=lambda x: x[1], reverse=True)
+        out_ids = [item[0] for item in sorted_dict_by_score[:step_size]]
 
         return out_ids
 
+    def compute_fluctuation_scores_per_doc(self, historical_sequences) -> Dict[int, float]:
+        fluctuation_scores: Dict[int, float] = dict()
+        for idx, hist_scores in historical_sequences.items():
+            fluctuation_scores[idx] = self.compute_fhs_score(hist_scores)
+        return fluctuation_scores
+
+    def compute_fhs_score(self, hist_scores: List[float]) -> float:
+        weight = self.weight  # in paper (11) referred as w_s
```

Comment:
Es muss glaube ich "referred to as w_s" heißen?

Reply:
You are right